### PR TITLE
Support more types in avro record decoder

### PR DESCRIPTION
**Pull request by elonazoulay**
This pull request extends the avro record reader:
* Support struct type
* Support nested list, map and struct types
* Support for float, integer, smallint and tinyint

### CODE REVIEW

**Comment by elonazoulay** [comment]
Added schema registry client to avro record decoder so that records with older schemas can be decoded.

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java [hunk 1]
```diff
@@ -80,7 +80,8 @@ private boolean isSupportedType(Type type)
 
         if (type instanceof ArrayType) {
             checkArgument(type.getTypeParameters().size() == 1, "expecting exactly one type parameter for array");
-            return isSupportedPrimitive(type.getTypeParameters().get(0));
+            return true;
+            //return isSupportedPrimitive(type.getTypeParameters().get(0));
```
Comment (findepi) [inline presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java]: What did you intend to do here? Recurse to `isSupportedType(..)`?

File: presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java [hunk 2]
```diff
@@ -448,6 +449,35 @@ public void testArrayDecodedAsArray()
         checkArrayValue(decodedRow, row, new long[] {114, 136});
     }
 
+    @Test
+    public void testNestedArrayDecodedAsArray()
```
Comment (findepi) [inline presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java]: Add a test with 3-level nesting.
Add a test with array of map.

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java (hunk 1 above)
Comment (kokosing) [inline presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java]: remove

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java [hunk 3]
```diff
@@ -81,14 +85,13 @@ private boolean isSupportedType(Type type)
...
             return true;
-            //return isSupportedPrimitive(type.getTypeParameters().get(0));
         }
 
         if (type instanceof MapType) {
             List<Type> typeParameters = type.getTypeParameters();
             checkArgument(typeParameters.size() == 2, "expecting exactly two type parameters for map");
             checkArgument(typeParameters.get(0) instanceof VarcharType, "Unsupported column type '%s' for map key", typeParameters.get(0));
-            return isSupportedPrimitive(type.getTypeParameters().get(1));
+            return true;
```
Comment (findepi) [inline presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java]: Shouldn't you recurse to `isSupportedType` here?

Also, separate this change out from `Add support for more primitive types` commit and add tests.

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java [hunk 4]
```diff
@@ -80,7 +80,8 @@ private boolean isSupportedType(Type type)
 
         if (type instanceof ArrayType) {
             checkArgument(type.getTypeParameters().size() == 1, "expecting exactly one type parameter for array");
-            return isSupportedPrimitive(type.getTypeParameters().get(0));
+            return true;
```
Comment (kokosing) [inline presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java]: do you support array of maps and other complex nested types in arrays?

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java [hunk 5]
```diff
@@ -201,41 +202,35 @@ private static Block serializeObject(BlockBuilder builder, Object value, Type ty
         if (type instanceof MapType) {
             return serializeMap(builder, value, type, columnName);
         }
-        serializeGeneric(builder, value, type, columnName);
+        serializePrimitive(builder, value, type, columnName);
         return null;
     }
 
-    private static Block serializeList(BlockBuilder blockBuilder, Object value, Type type, String columnName)
+    private static Block serializeList(BlockBuilder parentBlockBuilder, Object value, Type type, String columnName)
```
Comment (kokosing) [inline presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java]: undo rename or extract to separate commit

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java [hunk 6]
```diff
@@ -100,6 +103,10 @@ private boolean isSupportedPrimitive(Type type)
                         BigintType.BIGINT,
                         BooleanType.BOOLEAN,
                         DoubleType.DOUBLE,
+                        IntegerType.INTEGER,
+                        RealType.REAL,
+                        SmallintType.SMALLINT,
+                        TinyintType.TINYINT,
```
Comment (findepi) [inline presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java]: add integer types together, in order: tiny, small, int, biging

add real before double

make the list an ImmutableSet and make it a class constant

File: presto-kafka/src/test/java/io/prestosql/plugin/kafka/KafkaRawQueryRunner.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+import io.airlift.log.Logger;
+import io.airlift.log.Logging;
+import io.prestosql.Session;
+import io.prestosql.testing.DistributedQueryRunner;
+
+import java.util.Map;
+
+import static io.prestosql.testing.TestingSession.testSessionBuilder;
+
+public final class KafkaRawQueryRunner
```
Comment (kokosing) [inline presto-kafka/src/test/java/io/prestosql/plugin/kafka/KafkaRawQueryRunner.java]: why do you need another kafka query runner? Maybe the existing one should accept some parameters instead?

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java [hunk 8]
```diff
@@ -93,6 +96,10 @@ private boolean isSupportedType(Type type)
             checkArgument(typeParameters.get(0) instanceof VarcharType, "Unsupported column type '%s' for map key", typeParameters.get(0));
             return true;
         }
+
+        if (type instanceof RowType) {
```
Comment (kokosing) [inline presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java]: I wonder if you modify avro decoder if you don't need to check or add tests for `redis` connector that is also using this.

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroRowDecoder.java [hunk 9]
```diff
@@ -31,11 +34,16 @@
...
 import static java.util.Objects.requireNonNull;
+import static org.apache.avro.file.DataFileConstants.MAGIC;
 
 public class AvroRowDecoder
         implements RowDecoder
 {
     public static final String NAME = "avro";
+    public static final int HEADER_LENGTH = 1 + Integer.BYTES;
+
+    private static ThreadLocal<BinaryDecoder> reuseDecoder = ThreadLocal.withInitial(() -> null);
```
Comment (kokosing) [inline presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroRowDecoder.java]: Why do we need thread local? Is `BinaryDecoder` thread safe?

File: presto-kafka/src/test/java/io/prestosql/plugin/kafka/KafkaQueryRunner.java [hunk 10]
```diff
@@ -65,7 +65,9 @@ static DistributedQueryRunner createKafkaQueryRunner(TestingKafka testingKafka,
     {
```
Comment (kokosing) [inline presto-kafka/src/test/java/io/prestosql/plugin/kafka/KafkaQueryRunner.java]: this is going to be covered by https://github.com/prestosql/presto/pull/2164

Also commit message is wrong.`KafkaQueryRunner` is not deprecated.

File: pom.xml [hunk 11]
```diff
@@ -831,6 +831,22 @@
                 <version>1.9.1</version>
```
Comment (kokosing) [inline pom.xml]: you need tests for that. Please setup a docker container with kafka schema registry in tests and use it.

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java [hunk 12]
```diff
@@ -244,13 +251,13 @@ private static void serializePrimitive(BlockBuilder blockBuilder, Object value,
...
-            type.writeLong(blockBuilder, (Long) value);
+        if (type instanceof BigintType || type instanceof IntegerType || type instanceof SmallintType || type instanceof TinyintType) {
+            type.writeLong(blockBuilder, ((Number) value).longValue());
             return;
         }
 
-        if (type instanceof DoubleType) {
-            type.writeDouble(blockBuilder, (Double) value);
+        if (type instanceof DoubleType || type instanceof RealType) {
+            type.writeDouble(blockBuilder, ((Number) value).doubleValue());
```
Comment (findepi) [inline presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java]: for RealType you need to decode the value (int bits to float)

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java [hunk 13]
```diff
@@ -244,13 +251,13 @@ private static void serializePrimitive(BlockBuilder blockBuilder, Object value,
             return;
         }
 
-        if (type instanceof BigintType) {
-            type.writeLong(blockBuilder, (Long) value);
+        if (type instanceof BigintType || type instanceof IntegerType || type instanceof SmallintType || type instanceof TinyintType) {
+            type.writeLong(blockBuilder, ((Number) value).longValue());
```
Comment (findepi) [inline presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java]: `Number#longValue` is a code smell, as it can truncate/round the contained value

retain the cast to `Long`, since the other integer types are represented in Presto by a long too

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java (hunk 4 above)
Comment (elonazoulay) [reply presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java]: Yes, I will be adding more unit tests.

File: presto-kafka/src/test/java/io/prestosql/plugin/kafka/KafkaRawQueryRunner.java (hunk 7 above)
Comment (elonazoulay) [reply presto-kafka/src/test/java/io/prestosql/plugin/kafka/KafkaRawQueryRunner.java]: This should be removed, I was using it to test against a local kafka + schema registry with deeply nested avro types. I'll add tests.

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroRowDecoder.java (hunk 9 above)
Comment (elonazoulay) [reply presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroRowDecoder.java]: BinaryDecoder is not threadsafe. There is a configure() method which sets the data.

File: pom.xml (hunk 11 above)
Comment (elonazoulay) [reply pom.xml]: Absolutely! Will do.

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java (hunk 12 above)
Comment (elonazoulay) [reply presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java]: The value is coming from avro FLOAT, DOUBLE types so it would be a java Float or Double.

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java (hunk 13 above)
Comment (elonazoulay) [reply presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java]: Reverted: the value will either be a java Long or Integer as it is coming from avro LONG and INT. I will add unit tests for these as well.

Comment (elonazoulay) [reply presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java]: Had to modify this and add back the Number#longValue - verifying that the value is a java Integer or Long since that is what comes from avro.

Comment (findepi) [reply presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java]: > the value is a java Integer or Long since that is what comes from avro

i missed that. Okay, so use this information. it's really recommended to stay clear from `Number#longValue`

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java (hunk 1 above)
Comment (elonazoulay) [reply presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java]: Yes, now it recurses.

**Comment by kokosing** [comment]
@elonazoulay Is this ready for a second round of review? BTW please sort commits with: https://gist.github.com/findepi/7abf17c8a26d3b74e8bb9527caadfdbe

**Comment by elonazoulay** [comment]
Hey @kokosing! I would like to commit some more tests, just got the schema registry test container working. Also, I'll sort the commits. Should be ready shortly, I'll update as soon as it's ready. Thanks!

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java (hunk 12 above)
Comment (elonazoulay) [reply presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroColumnDecoder.java]: Yep, I was wrong, separated the Double and RealType conversion. Added a test to verify.

File: presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java [hunk 14]
```diff
@@ -523,7 +548,219 @@ public void testArrayDecodedAsArray()
...
+    @Test
+    public void testNestedLongArrayWithNulls()
+    {
+        DecoderTestColumnHandle row = new DecoderTestColumnHandle(0, "row", new ArrayType(new ArrayType(BIGINT)), "array_field", null, null, false, false, false);
+        Schema schema = SchemaBuilder.array().items().nullable().array().items().nullable().longType();
+        List<List<Long>> data = Stream.of(
+                ImmutableList.of(12L, 15L, 17L),
+                ImmutableList.of(22L, 25L, 27L, 29L),
+                null,
+                Stream.of(3L, 5L, null, 6L).collect(toList()))
```
Comment (kokosing) [inline presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java]: `Arrays.asList()`?

File: presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java [hunk 15]
```diff
@@ -523,7 +548,219 @@ public void testArrayDecodedAsArray()
...
+
+        checkArrayValue(decodedRow, row, list);
+    }
+
+    @Test
+    public void testNestedLongArrayWithNulls()
+    {
+        DecoderTestColumnHandle row = new DecoderTestColumnHandle(0, "row", new ArrayType(new ArrayType(BIGINT)), "array_field", null, null, false, false, false);
+        Schema schema = SchemaBuilder.array().items().nullable().array().items().nullable().longType();
+        List<List<Long>> data = Stream.of(
```
Comment (kokosing) [inline presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java]: `Arrays.asList()` instead of `Stream.of`?

File: presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java [hunk 16]
```diff
@@ -523,7 +548,219 @@ public void testArrayDecodedAsArray()
...
+
+        checkArrayValue(decodedRow, row, list);
+    }
+
+    @Test
+    public void testNestedStringArrayWithNulls()
+    {
+        DecoderTestColumnHandle row = new DecoderTestColumnHandle(0, "row", new ArrayType(new ArrayType(VARCHAR)), "array_field", null, null, false, false, false);
+        Schema schema = SchemaBuilder.array().items().nullable().array().items().nullable().stringType();
+        List<List<String>> data = Stream.of(
```
Comment (kokosing) [inline presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java]: same comments about `Arrays.asList()`

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroRowDecoder.java [hunk 17]
```diff
@@ -55,7 +63,59 @@ private AvroColumnDecoder createColumnDecoder(DecoderColumnHandle columnHandle)
...
+        GenericRecord avroRecord = candidate
+                        .orElseThrow(() -> new PrestoException(GENERIC_INTERNAL_ERROR, "Decoding Avro record failed."));
+
+        return Optional.of(columnDecoders.entrySet().stream()
+                .collect(toImmutableMap(
+                        Map.Entry::getKey,
+                        entry -> entry.getValue().decodeField(avroRecord))));
+    }
+
+    private void closeQuietly(DataFileStream<GenericRecord> stream)
```
Comment (kokosing) [inline presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroRowDecoder.java]: why not to close loudly? Why you swallow exception?

File: pom.xml [hunk 18]
```diff
@@ -831,6 +831,22 @@
...
             </dependency>
 
+            <dependency>
+                <groupId>io.confluent</groupId>
```
Comment (kokosing) [inline pom.xml]: What is the licence of this?

File: pom.xml [hunk 19]
```diff
@@ -831,6 +831,22 @@
...
+                <artifactId>kafka-schema-registry-client</artifactId>
+                <version>5.3.1</version>
```
Comment (kokosing) [inline pom.xml]: IIRC we are using 5.2.1 in testing for now.

File: presto-record-decoder/pom.xml [hunk 20]
```diff
@@ -56,6 +56,11 @@
             <artifactId>avro</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>io.confluent</groupId>
```
Comment (kokosing) [inline presto-record-decoder/pom.xml]: Is schema-registry only supported with confluent? Is there any community version of schema registry that comes with apache kafka?

File: presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroRowDecoderFactory.java [hunk 21]
```diff
@@ -13,25 +13,36 @@
  */
 package io.prestosql.decoder.avro;
 
+import io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient;
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
```
Comment (kokosing) [inline presto-record-decoder/src/main/java/io/prestosql/decoder/avro/AvroRowDecoderFactory.java]: I would prefer not to use kafka specific code in generic `presto-record-decoder`. I would prefer to introduce some extensions points to `AvroRowDecoderFactory` and use them in `presto-kafka` module to use kafka schema registry there.

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaMetadata.java [hunk 22]
```diff
@@ -48,19 +58,30 @@
...
+            TypeManager typeManager)
     {
         requireNonNull(kafkaConfig, "kafkaConfig is null");
         this.hideInternalColumns = kafkaConfig.isHideInternalColumns();
 
         requireNonNull(kafkaTableDescriptionSupplier, "kafkaTableDescriptionSupplier is null");
         this.tableDescriptions = kafkaTableDescriptionSupplier.get();
+
+        requireNonNull(typeManager, "typeManager is null");
+        this.schemaRegistryClient = kafkaConfig.getSchemaRegistryUrl().map(schemaRegistryUrl ->
```
Comment (kokosing) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaMetadata.java]: no need to wrap line, or wrap line before `.map`

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaSplitManager.java [hunk 23]
```diff
@@ -111,7 +121,35 @@ private static TopicPartition toTopicPartition(PartitionInfo partitionInfo)
...
+    {
+        if (!isAutoResolvable(tableHandle, isKey)) {
+            return Optional.empty();
+        }
+        try {
+            SchemaMetadata schemaMetadata = schemaRegistryClient.get().getLatestSchemaMetadata(tableHandle.getTopicName() + getSuffix(isKey));
+            return Optional.of((new Schema.Parser()).parse(schemaMetadata.getSchema()).toString());
+        }
+        catch (IOException | RestClientException e) {
+            throw new TableNotFoundException(tableHandle.toSchemaTableName());
```
Comment (kokosing) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaSplitManager.java]: IOException nor RestClientException does not mean that table does not exists, doesn't it?

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaMetadata.java [hunk 24]
```diff
@@ -223,6 +260,89 @@ private ConnectorTableMetadata getTableMetadata(SchemaTableName schemaTableName)
...
+    {
+        if (isKey) {
+            return tableHandle.getKeyDataSchemaLocation().map(KafkaMetadata::shouldAutoResolveSchema).orElse(false)
+                    && tableHandle.getKeyDataFormat().equalsIgnoreCase(AVRO);
+        }
+        else {
+            return tableHandle.getMessageDataSchemaLocation().map(KafkaMetadata::shouldAutoResolveSchema).orElse(false)
+                    && tableHandle.getMessageDataFormat().equalsIgnoreCase(AVRO);
+        }
+    }
```
Comment (kokosing) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaMetadata.java]: rewrite this to:
```
  static boolean isAutoResolvable(KafkaTableHandle tableHandle, boolean isKey)
    {
        return getLocation(tableHandle, isKey).map(KafkaMetadata::shouldAutoResolveSchema).orElse(false)
                && tableHandle.getKeyDataFormat().equalsIgnoreCase(AVRO);
    }

    private static Optional<String> getLocation(KafkaTableHandle tableHandle, boolean isKey)
    {
        if (isKey) {
            return tableHandle.getKeyDataSchemaLocation();
        }
        return tableHandle.getMessageDataSchemaLocation();
}
```

You can apply same for `isAutoResolvable`

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaMetadata.java [hunk 25]
```diff
@@ -16,6 +16,11 @@
 import com.google.common.collect.ImmutableList;
```
Comment (kokosing) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaMetadata.java]: I don't see unit tests that should go in this commit.

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaMetadata.java [hunk 26]
```diff
@@ -48,19 +58,30 @@
 public class KafkaMetadata
         implements ConnectorMetadata
 {
+    private static final String AUTO_RESOLVE = "auto";
+    private static final String AVRO = "avro";
+
     private final boolean hideInternalColumns;
     private final Map<SchemaTableName, KafkaTopicDescription> tableDescriptions;
+    private final Optional<SchemaRegistryClient> schemaRegistryClient;
+    private final AvroSchemaConverter avroSchemaConverter;
```
Comment (kokosing) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaMetadata.java]: Schema discovery is becoming quite complicated logic now. I would extract this out of `KafkaMetadata` and made it clear when the schema is read from where.

Also, is it possible to not declare tables at all in `tableDescriptions` so kafka connector could discovers topics automatically? Thanks to that it would be much easier to configure kafka connector.

Comment (elonazoulay) [reply presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaMetadata.java]: Yes it is! We have code that reads schema from schema-registry without json files, will add that as well.

File: pom.xml (hunk 18 above)
Comment (elonazoulay) [reply pom.xml]: Apache License 2.0:
https://github.com/confluentinc/schema-registry/blob/master/client/src/main/java/io/confluent/kafka/schemaregistry/client/SchemaRegistryClient.java

File: presto-record-decoder/pom.xml (hunk 20 above)
Comment (elonazoulay) [reply presto-record-decoder/pom.xml]: There are other types of schema registries but they have different api's and are not distributed with apache kafka , this one is only supported with confluent, it's the most widely used one.

**Comment by elonazoulay** [comment]
Split into 2 pull requests. The schema registry pull request is #2499

**Comment by kokosing** [comment]
@findepi do you want to take a look?

**Comment by kokosing** [comment]
@aalbu and @charlesjmorgan Do you want to review?

@elonazoulay Can you please rebase? I would like to merge it.

**Comment by kokosing** [comment]
Merged, thanks! Sorry for delay.
